presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/FileStorageService.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    @PreDestroy
+    public void stop()
+            throws IOException
+    {
+        deleteDirectory(baseStagingDir);
+    }
+
+    @Override
+    public File getStorageFile(long chunkId)
+    {
+        return createParents(getFileSystemPath(baseStorageDir, chunkId));
+    }
+
+    @Override
+    public File getStagingFile(long chunkId)
+    {
+        String name = getFileSystemPath(new File("/"), chunkId).getName();
+        return createParents(new File(baseStagingDir, name));
+    }
+
+    @Override
+    public File getQuarantineFile(long chunkId)
+    {
+        String name = getFileSystemPath(new File("/"), chunkId).getName();
+        return createParents(new File(baseQuarantineDir, name));
+    }
+
+    @Override
+    public Set<Long> getStorageChunks()
+    {
+        ImmutableSet.Builder<Long> chunks = ImmutableSet.builder();
+        for (File level1 : listFiles(baseStorageDir, FileStorageService::isHexDirectory)) {
+            for (File level2 : listFiles(level1, FileStorageService::isHexDirectory)) {
+                for (File file : listFiles(level2, path -> true)) {
+                    if (file.isFile()) {
+                        chunkIdFromFileName(file.getName()).ifPresent(chunks::add);
+                    }
+                }
+            }
+        }
+        return chunks.build();
+    }
+
+    /**
+     * Generate a file system path for a chunk ID.
+     * This creates a three level deep directory structure where the first
+     * two levels each contain two hex digits (lowercase) of the xxhash64
+     * of the ID in big endian hexadecimal and the final level contains the
+     * full ID in decimal. Example:
+     * <pre>
+     * ID: 12345678
+     * Hash: e4fc629b8375452c
+     * Path: /base/e4/fc/12345678
+     * </pre>
+     * This ensures that files are spread out evenly through the tree
+     * while a path can still be easily navigated by a human being.
+     */
+    public static File getFileSystemPath(File base, long chunkId)
+    {
+        checkArgument(chunkId >= 0, "chunkId is negative");
+        String hash = format("%08x", XxHash64.hash(chunkId));
+        return base.toPath()
+                .resolve(hash.substring(0, 2))
+                .resolve(hash.substring(2, 4))
+                .resolve(String.valueOf(chunkId))
+                .toFile();
+    }
+
+    private static File createParents(File file)
+    {
+        try {
+            createParentDirectories(file.toPath());
+            return file;
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed creating parent directories: " + file, e);
+        }
+    }
+
+    private static void createDirectory(File file)
+    {
+        try {
+            createDirectories(file.toPath());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Failed to create directory: " + file, e);
+        }
+    }
+
+    private void deleteStagingFilesAsync()
```

Comment:
"highker on Feb 4  Member
Why making this async given getStagingFile can create new staging files while we are still deleting old staging files?

  @zhaotao1986
zhaotao1986 on Feb 14  Member
FileName, either StagingFIle or StorageFile is just ChunkID; ChunkID doesn't repeat. So this thread won't interfere getStagingFile(). So sync or async should be considered in other aspects.

Other aspects like, where to put blocking operations during system start. I gathered in quip."